qa/tasks/caasp.py

```diff
@@ -0,0 +1,137 @@
+'''
+Task that deploys a CAASP cluster on all the nodes
+Linter:
+    flake8 --max-line-length=100
+'''
+import logging
+import os
+import subprocess
+from util import remote_exec
+from teuthology.exceptions import ConfigError
+from teuthology.misc import (
+    delete_file,
+    move_file,
+    sh,
+    sudo_write_file,
+    write_file,
+    copy_file,
+    all_roles_of_type
+    )
+from teuthology.orchestra import run
+from teuthology.task import Task
+from util import (
+    get_remote_for_role,
+    remote_exec
+    )
+log = logging.getLogger(__name__)
+
+
+class Caasp(Task):
+    """
+    Deploy a Salt cluster on all remotes (test nodes).
+    This task assumes all relevant Salt packages (salt, salt-master,
+    salt-minion, salt-api, python-salt, etc. - whatever they may be called for
+    the OS in question) are already installed. This should be done using the
+    install task.
+    One, and only one, of the machines must have a role corresponding to the
+    value of the variable salt.sm.master_role (see salt_manager.py). This node
+    is referred to as the "Salt Master", or the "master node".
+    The task starts the Salt Master daemon on the master node, and Salt Minion
+    daemons on all the nodes (including the master node), and ensures that the
+    minions are properly linked to the master. Finally, it tries to ping all
+    the minions from the Salt Master.
+    :param ctx: the argparse.Namespace object
+    :param config: the config dict
+    """
+
+    def __init__(self, ctx, config):
+        super(Caasp, self).__init__(ctx, config)
+        log.debug("beginning of constructor method")
+        self.ctx['roles'] = self.ctx.config['roles']
+        self.log = log
+        self.remotes = self.cluster.remotes
+        self.mgmt_remote = get_remote_for_role(self.ctx, "skuba_mgmt_host.0")
+        self.ssh_priv = 'tempkey.rsa'
+        self.ssh_pub = 'tempkey.rsa.pub'
+        self.set_agent = "eval `ssh-agent` && ssh-add ~/.ssh/%s && " % self.ssh_priv
+
+    def __ssh_setup(self):
+        log.debug("Executing SSH setup")
+        self.__ssh_gen_key()
+        self.__ssh_copy_priv()
+        self.__ssh_copy_pub_to_caasp()
+
+    def __ssh_gen_key(self):
+        if os.path.isfile(self.ssh_priv):
+            os.remove(self.ssh_priv)
+        if os.path.isfile(self.ssh_pub):
+            os.remove(self.ssh_pub)
+        os.system('ssh-keygen -t rsa -b 2048 -P "" -f %s' % self.ssh_priv)
```

Comment:
This code is running on the worker, and as I understand it is not thread safe, we should to avoid situation kind of that. Also, as far as I can see, Nathan's variant makes sense to me. Especially that the remote nodes can have different ssh config from the teuthology worker. You can generate the keys remotely, capture key contents to corresponding variables and write the files remotely.